{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1393bd78",
   "metadata": {},
   "source": [
    "# Homework ‚Ññ1\n",
    "\n",
    "Load reads from all experiments from article¬†**[Evaluation of two main RNA-seq approaches for gene quantification in clinical RNA sequencing: polyA+ selection versus rRNA depletion](https://www.nature.com/articles/s41598-018-23226-4)**. \n",
    "\n",
    "In this article authors compared two methods of removing highly abundant ribosomal RNAs (rRNA): positive polyA+ selection and rRNA depletion (negative selection). They evaluated these two RNA sequencing approaches using human blood and colon tissue samples.\n",
    "\n",
    "Use the Nextflow pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ef01c9",
   "metadata": {},
   "source": [
    "## Let's start!\n",
    "1. Where can I find data that they used in this study?\n",
    "\n",
    "> All the raw sequencing reads generated in this study have been submitted to the NCBI Sequence Read Archive and are available under the accession number **SRP127360**.\n",
    ">\n",
    "\n",
    "2. Download file with a list of accessions: \n",
    "\n",
    "Go to NCBI ‚Üí Choose ‚ÄúSRA‚Äù as Database and look for ‚ÄúSRP127360‚Äù ‚Üí Click ‚ÄúSend to‚Äù ‚Üí Choose Destination - ‚ÄúFile‚Äù ‚Üí Format - ‚ÄúAccession List‚Äù\n",
    "After that you will obtain file called SraAccList.txt with 16 accession numbers.\n",
    "\n",
    "3. Install needed tools and programs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1681aa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install SRA toolkit \n",
    "! curl --output sratoolkit.tar.gz https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-mac64.tar.gz\n",
    "! tar -vxzf sratoolkit.tar.gz\n",
    "! vdb-config -i \n",
    "\n",
    "# Install FastQC\n",
    "! curl --output fastqc_v0.11.9.zip https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
    "! unzip fastqc_v0.11.9.zip\n",
    "! chmod +x FastQC/fastqc\n",
    "\n",
    "# Install MultyQC\n",
    "! pip3 install multiqc\n",
    "\n",
    "# Install NextFlow\n",
    "! curl -fsSL https://get.nextflow.io | bash\n",
    "    \n",
    "# For opening .dot files\n",
    "brew install graphviz "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "291ed0fe",
   "metadata": {},
   "source": [
    "## Versions of programs and OS version of local machine\n",
    "* SRA toolkit: sratoolkit.3.0.0-mac64\n",
    "* FastQC: fastqc_v0.11.9 \n",
    "* MultiQC: version 1.13 (6e2b162)\n",
    "* Nextflow: version 22.04.5 build 5708\n",
    "* MacOS Version:\t12.3.1\n",
    "* MacOS BuildVersion:\t21E258"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "07830b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add path to programs in ${PATH}\n",
    "! export PATH=$PATH:/Users/mar4/transcriptomics/software/sratoolkit.3.0.0-mac64/bin\n",
    "! export PATH=$PATH:/Users/mar4/transcriptomics/software/FastQC\n",
    "! export PATH=$PATH:/Users/mar4/transcriptomics/software/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbe95a4d",
   "metadata": {},
   "source": [
    "## Test run for two accessions from lecture\n",
    "Every downloaded tool has run succesfully. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9aef5c6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spots read      : 135,014\n",
      "reads read      : 270,028\n",
      "reads written   : 270,028\n",
      "spots read      : 158,874\n",
      "reads read      : 317,748\n",
      "reads written   : 317,748\n",
      "Started analysis of SRR20851170_1.fastq.gz\n",
      "Approx 5% complete for SRR20851170_1.fastq.gz\n",
      "Approx 10% complete for SRR20851170_1.fastq.gz\n",
      "Approx 15% complete for SRR20851170_1.fastq.gz\n",
      "Approx 20% complete for SRR20851170_1.fastq.gz\n",
      "Approx 25% complete for SRR20851170_1.fastq.gz\n",
      "Approx 30% complete for SRR20851170_1.fastq.gz\n",
      "Approx 35% complete for SRR20851170_1.fastq.gz\n",
      "Approx 40% complete for SRR20851170_1.fastq.gz\n",
      "Approx 45% complete for SRR20851170_1.fastq.gz\n",
      "Approx 50% complete for SRR20851170_1.fastq.gz\n",
      "Approx 55% complete for SRR20851170_1.fastq.gz\n",
      "Approx 60% complete for SRR20851170_1.fastq.gz\n",
      "Approx 65% complete for SRR20851170_1.fastq.gz\n",
      "Approx 70% complete for SRR20851170_1.fastq.gz\n",
      "Approx 75% complete for SRR20851170_1.fastq.gz\n",
      "Approx 80% complete for SRR20851170_1.fastq.gz\n",
      "Approx 85% complete for SRR20851170_1.fastq.gz\n",
      "Approx 90% complete for SRR20851170_1.fastq.gz\n",
      "Approx 95% complete for SRR20851170_1.fastq.gz\n",
      "Analysis complete for SRR20851170_1.fastq.gz\n",
      "Started analysis of SRR20851170_2.fastq.gz\n",
      "Approx 5% complete for SRR20851170_2.fastq.gz\n",
      "Approx 10% complete for SRR20851170_2.fastq.gz\n",
      "Approx 15% complete for SRR20851170_2.fastq.gz\n",
      "Approx 20% complete for SRR20851170_2.fastq.gz\n",
      "Approx 25% complete for SRR20851170_2.fastq.gz\n",
      "Approx 30% complete for SRR20851170_2.fastq.gz\n",
      "Approx 35% complete for SRR20851170_2.fastq.gz\n",
      "Approx 40% complete for SRR20851170_2.fastq.gz\n",
      "Approx 45% complete for SRR20851170_2.fastq.gz\n",
      "Approx 50% complete for SRR20851170_2.fastq.gz\n",
      "Approx 55% complete for SRR20851170_2.fastq.gz\n",
      "Approx 60% complete for SRR20851170_2.fastq.gz\n",
      "Approx 65% complete for SRR20851170_2.fastq.gz\n",
      "Approx 70% complete for SRR20851170_2.fastq.gz\n",
      "Approx 75% complete for SRR20851170_2.fastq.gz\n",
      "Approx 80% complete for SRR20851170_2.fastq.gz\n",
      "Approx 85% complete for SRR20851170_2.fastq.gz\n",
      "Approx 90% complete for SRR20851170_2.fastq.gz\n",
      "Approx 95% complete for SRR20851170_2.fastq.gz\n",
      "Analysis complete for SRR20851170_2.fastq.gz\n",
      "Started analysis of SRR20851171_1.fastq.gz\n",
      "Approx 5% complete for SRR20851171_1.fastq.gz\n",
      "Approx 10% complete for SRR20851171_1.fastq.gz\n",
      "Approx 15% complete for SRR20851171_1.fastq.gz\n",
      "Approx 20% complete for SRR20851171_1.fastq.gz\n",
      "Approx 25% complete for SRR20851171_1.fastq.gz\n",
      "Approx 30% complete for SRR20851171_1.fastq.gz\n",
      "Approx 35% complete for SRR20851171_1.fastq.gz\n",
      "Approx 40% complete for SRR20851171_1.fastq.gz\n",
      "Approx 45% complete for SRR20851171_1.fastq.gz\n",
      "Approx 50% complete for SRR20851171_1.fastq.gz\n",
      "Approx 55% complete for SRR20851171_1.fastq.gz\n",
      "Approx 60% complete for SRR20851171_1.fastq.gz\n",
      "Approx 65% complete for SRR20851171_1.fastq.gz\n",
      "Approx 70% complete for SRR20851171_1.fastq.gz\n",
      "Approx 75% complete for SRR20851171_1.fastq.gz\n",
      "Approx 80% complete for SRR20851171_1.fastq.gz\n",
      "Approx 85% complete for SRR20851171_1.fastq.gz\n",
      "Approx 90% complete for SRR20851171_1.fastq.gz\n",
      "Approx 95% complete for SRR20851171_1.fastq.gz\n",
      "Analysis complete for SRR20851171_1.fastq.gz\n",
      "Started analysis of SRR20851171_2.fastq.gz\n",
      "Approx 5% complete for SRR20851171_2.fastq.gz\n",
      "Approx 10% complete for SRR20851171_2.fastq.gz\n",
      "Approx 15% complete for SRR20851171_2.fastq.gz\n",
      "Approx 20% complete for SRR20851171_2.fastq.gz\n",
      "Approx 25% complete for SRR20851171_2.fastq.gz\n",
      "Approx 30% complete for SRR20851171_2.fastq.gz\n",
      "Approx 35% complete for SRR20851171_2.fastq.gz\n",
      "Approx 40% complete for SRR20851171_2.fastq.gz\n",
      "Approx 45% complete for SRR20851171_2.fastq.gz\n",
      "Approx 50% complete for SRR20851171_2.fastq.gz\n",
      "Approx 55% complete for SRR20851171_2.fastq.gz\n",
      "Approx 60% complete for SRR20851171_2.fastq.gz\n",
      "Approx 65% complete for SRR20851171_2.fastq.gz\n",
      "Approx 70% complete for SRR20851171_2.fastq.gz\n",
      "Approx 75% complete for SRR20851171_2.fastq.gz\n",
      "Approx 80% complete for SRR20851171_2.fastq.gz\n",
      "Approx 85% complete for SRR20851171_2.fastq.gz\n",
      "Approx 90% complete for SRR20851171_2.fastq.gz\n",
      "Approx 95% complete for SRR20851171_2.fastq.gz\n",
      "Analysis complete for SRR20851171_2.fastq.gz\n",
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=11549;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ üîç \u001b[2m| v1.13 (6e2b162)\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /Users/mar4/transcriptomics/homework/notebooks\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m14/14\u001b[0m  \u001b[0m \u001b[2m./SRR20851170_1_fastqc.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 4 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "# Download two files \n",
    "! fasterq-dump --split-files SRR20851170\n",
    "! fasterq-dump --split-files SRR20851171\n",
    "\n",
    "# Compress four files\n",
    "! bgzip SRR20851170_1.fastq\n",
    "! bgzip SRR20851170_2.fastq\n",
    "! bgzip SRR20851171_1.fastq\n",
    "! bgzip SRR20851171_2.fastq\n",
    "\n",
    "# Run FastQC (We can use all files as input)\n",
    "! fastqc SRR20851170_1.fastq.gz SRR20851170_2.fastq.gz SRR20851171_1.fastq.gz SRR20851171_2.fastq.gz\n",
    "\n",
    "# Run MultiQC (We should give directory with files after running FastQC)\n",
    "! multiqc ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06def9bb",
   "metadata": {},
   "source": [
    "## Run Nextflow script and check the results\n",
    "There are 16 accessions. It is too much for my computer, so I decided to run pipeline on two accession: one sample is from colon (SRR6410610) and the other is from blood (SRR6410618).\n",
    "\n",
    "As input nextflow pipeline use file SraAccList.txt where every line is separated accession number.\n",
    "Nextflow script download FastQ files with fasterq-dump, then evaluate quality with FastQC and join its results with MultiQC. Output is MultiQC html report.\n",
    "\n",
    "**Flags:**<br />\n",
    "-with-report - create processes execution html report.<br />\n",
    "-with-dag - create pipeline DAG file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8cd26e31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N E X T F L O W  ~  version 22.04.5\n",
      "Launching `hw1_intro.nf` [infallible_mendel] DSL2 - revision: 2eb2a7558d\n",
      "\n",
      "  Q U A L I T Y   C O N T R O L  \n",
      "=================================\n",
      "SRA numbers in file        : SraAccList.txt\n",
      "Results location   : results/multiqc_report/\n",
      "executor >  local (2)\u001b[K\n",
      "[32/38ed37] process > DownloadFastQ (1) [  0%] 0 of 2\u001b[K\n",
      "[-        ] process > RunFastQC         -\u001b[K\n",
      "[-        ] process > RunMultiQC        -\u001b[K\n",
      "\u001b[33mWARN: Task runtime metrics are not reported when using macOS without a container engine\u001b[39m\u001b[K\n",
      "\u001b[6A\n",
      "executor >  local (2)\u001b[K\n",
      "[32/38ed37] process > DownloadFastQ (1) [  0%] 0 of 2\u001b[K\n",
      "[-        ] process > RunFastQC         -\u001b[K\n",
      "[-        ] process > RunMultiQC        -\u001b[K\n",
      "\u001b[33mWARN: Task runtime metrics are not reported when using macOS without a container engine\u001b[39m\u001b[K\n",
      "\u001b[6A\n",
      "executor >  local (3)\u001b[K\n",
      "[89/496d2f] process > DownloadFastQ (2) [ 50%] 1 of 2\u001b[K\n",
      "[d1/ea5379] process > RunFastQC (1)     [  0%] 0 of 1\u001b[K\n",
      "[-        ] process > RunMultiQC        -\u001b[K\n",
      "\u001b[33mWARN: Task runtime metrics are not reported when using macOS without a container engine\u001b[39m\u001b[K\n",
      "\u001b[K\n",
      "\u001b[7A\n",
      "executor >  local (4)\u001b[K\n",
      "[32/38ed37] process > DownloadFastQ (1) [100%] 2 of 2 ‚úî\u001b[K\n",
      "[70/9726be] process > RunFastQC (2)     [  0%] 0 of 2\u001b[K\n",
      "[-        ] process > RunMultiQC        -\u001b[K\n",
      "\u001b[33mWARN: Task runtime metrics are not reported when using macOS without a container engine\u001b[39m\u001b[K\n",
      "\u001b[K\n",
      "\u001b[7A\n",
      "executor >  local (4)\u001b[K\n",
      "[32/38ed37] process > DownloadFastQ (1) [100%] 2 of 2 ‚úî\u001b[K\n",
      "[d1/ea5379] process > RunFastQC (1)     [ 50%] 1 of 2\u001b[K\n",
      "[-        ] process > RunMultiQC        -\u001b[K\n",
      "\u001b[K\n",
      "\u001b[6A\n",
      "executor >  local (5)\u001b[K\n",
      "[32/38ed37] process > DownloadFastQ (1) [100%] 2 of 2 ‚úî\u001b[K\n",
      "[70/9726be] process > RunFastQC (2)     [100%] 2 of 2 ‚úî\u001b[K\n",
      "[3e/1a820e] process > RunMultiQC        [  0%] 0 of 1\u001b[K\n",
      "\u001b[33mWARN: Task runtime metrics are not reported when using macOS without a container engine\u001b[39m\u001b[K\n",
      "\u001b[6A\n",
      "executor >  local (5)\u001b[K\n",
      "[32/38ed37] process > DownloadFastQ (1) [100%] 2 of 2 ‚úî\u001b[K\n",
      "[70/9726be] process > RunFastQC (2)     [100%] 2 of 2 ‚úî\u001b[K\n",
      "[3e/1a820e] process > RunMultiQC        [100%] 1 of 1 ‚úî\u001b[K\n",
      "\u001b[33mWARN: Task runtime metrics are not reported when using macOS without a container engine\u001b[39m\u001b[K\n",
      "\u001b[6A\n",
      "executor >  local (5)\u001b[K\n",
      "[32/38ed37] process > DownloadFastQ (1) [100%] 2 of 2 ‚úî\u001b[K\n",
      "[70/9726be] process > RunFastQC (2)     [100%] 2 of 2 ‚úî\u001b[K\n",
      "[3e/1a820e] process > RunMultiQC        [100%] 1 of 1 ‚úî\u001b[K\n",
      "\u001b[33mWARN: Task runtime metrics are not reported when using macOS without a container engine\u001b[39m\u001b[K\n",
      "\u001b[32;1mCompleted at: 23-Sep-2022 03:30:18\n",
      "Duration    : 2h 35m 15s\n",
      "CPU hours   : 5.2\n",
      "Succeeded   : 5\n",
      "\u001b[22;39m\u001b[K\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! nextflow run hw1_intro.nf -with-report results/nextflow_report.html -with-dag results/flowchart.pdf --SRA_acc_list 'SraAccList.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f3e8ee",
   "metadata": {},
   "source": [
    "Also we can look at hw1_intro.nf!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9a1e0170",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/usr/bin/env nextflow\r\n",
      "\r\n",
      "params.SRA_acc_list = \"filename.txt\"\r\n",
      "params.results_dir = \"results/multiqc_report/\"\r\n",
      "\r\n",
      "log.info \"\"\r\n",
      "log.info \"  Q U A L I T Y   C O N T R O L  \"\r\n",
      "log.info \"=================================\"\r\n",
      "log.info \"SRA numbers in file        : ${params.SRA_acc_list}\"\r\n",
      "log.info \"Results location   : ${params.results_dir}\"\r\n",
      "\r\n",
      "File acc_list = new File(params.SRA_acc_list)\r\n",
      "SRA_list = acc_list.getText('UTF-8').split(\"\\n\")\r\n",
      "\r\n",
      "process DownloadFastQ {\r\n",
      "  input:\r\n",
      "    val sra_acc\r\n",
      "\r\n",
      "  output:\r\n",
      "    path \"${sra_acc}/*\"\r\n",
      "\r\n",
      "  script:\r\n",
      "    \"\"\"\r\n",
      "    fasterq-dump --split-files -O ${sra_acc}/ ${sra_acc} \r\n",
      "    gzip -r ${sra_acc}/\r\n",
      "    \"\"\"\r\n",
      "}\r\n",
      "\r\n",
      "process RunFastQC {\r\n",
      "  input:\r\n",
      "    path fastq_files\r\n",
      "\r\n",
      "  output:\r\n",
      "    path \"qc/*\"\r\n",
      "\r\n",
      "  script:\r\n",
      "    \"\"\"\r\n",
      "    mkdir qc\r\n",
      "    fastqc -o qc $fastq_files\r\n",
      "    \"\"\"\r\n",
      "}\r\n",
      "\r\n",
      "process RunMultiQC {\r\n",
      "publishDir \"${params.results_dir}\"\r\n",
      "  input:\r\n",
      "    path fastqc_out\r\n",
      "  output:\r\n",
      "    path \"multiqc_report.html\"\r\n",
      "  script:\r\n",
      "    \"\"\"\r\n",
      "    multiqc ${fastqc_out}\r\n",
      "    \"\"\"\r\n",
      "}\r\n",
      "\r\n",
      "workflow {\r\n",
      "  data = Channel.of( SRA_list )\r\n",
      "  DownloadFastQ(data)\r\n",
      "  RunFastQC( DownloadFastQ.out )\r\n",
      "  RunMultiQC( RunFastQC.out.collect() )\r\n",
      "}\r\n"
     ]
    }
   ],
   "source": [
    "! cat hw1_intro.nf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
